File: Hamming.java
/**
 * @author Liby Lee
 * @date 09-26-2016
 */

/**
 * Hamming Class, Provide apis for encoding and decoding 4 digit codes
 */
public class Hamming {
    
    //What digit the parity bit covers
    private static int[][] parityCoversDigit = {{0, 1, 3},  //p1 covers d1, d2, d4
                                                {0, 2, 3},  //p2 covers d1, d3, d4
                                                {1, 2, 3}}; //p3 covers d2, d3, d4
    
    /**
     * Encode the length 4 data
     * @param data the data to be encoded
     * @return the encoded int array data
     */
    public static int[] encode(int[] data) {
        int[] parity = getParity(data); //p1, p2 and p3
        
        //Note: append a 0 in the end of the array
        int[] encoded = {parity[0], parity[1], data[0], parity[2], data[1], data[2], data[3], 0};
        
        //Calculate the num of the '1's. And append the extra parity bit to the encoded data
        int sum = 0;
        for (int i = 0; i < encoded.length; i++) {
            sum += encoded[i];
        }
        encoded[7] = sum % 2;
        return encoded;
    }
    
    /**
     * Decode the encoded code
     * @param encoded the encoded length 8 array.
     * @return int[] the decoded length 4 data.
     * @throws InvalidArgumentException if there's 2 errors in the encoded data
     */
    public static int[] decode(int[] encoded) {
        
        //Save all the useful data
        int[] data = {encoded[2], encoded[4], encoded[5], encoded[6]};
        int[] parity = {encoded[0], encoded[1], encoded[3]};
        
        //Generate the expected correctParity based on the 4 digit data
        int[] correctParity = getParity(data);
        
        //Compare the given parity and the correct parity
        //If like original p1 equals expected p1, then comparison[0] is true
        boolean[] comparison = {false, false, false};
        for (int i = 0; i < comparison.length; i++) {
            comparison[i] = parity[i] == correctParity[i];
        }
        
        //Check if there's an error in the encoded array
        //And save that position in the incorrectPosition
        int incorrectPosition = -1;
        if (!comparison[0] && comparison[1] && comparison[2]) { //p1 wrong
            incorrectPosition = 0;
        }
        if (comparison[0] && !comparison[1] && comparison[2]) { //p2 wrong
            incorrectPosition = 1;
        }
        if (!comparison[0] && !comparison[1] && comparison[2]) { //p1, p2 wrong
            incorrectPosition = 2;
        }
        if (comparison[0] && comparison[1] && !comparison[2]) { //p3 wrong
            incorrectPosition = 3;
        }
        if (!comparison[0] && comparison[1] && !comparison[2]) { //p1, p3 wrong
            incorrectPosition = 4;
        }
        if (comparison[0] && !comparison[1] && !comparison[2]) { //p2, p3 wrong
            incorrectPosition = 5;
        }
        if (!comparison[0] && !comparison[1] && !comparison[2]) { //all three are wrong
            incorrectPosition = 6;
        }
        
        //Check if there's an incorrect digit
        if (incorrectPosition != -1) {
            
            //Flip the digit on the [incorrectPosition] of the 8 digit encoded code
            encoded[incorrectPosition] = encoded[incorrectPosition] == 0 ? 1 : 0;
            
            //Ok we now flipped the digit.
            //Then we need to calculate how many 1's are in the encoded array.
            //This will be stored in the [sum]
            int sum = 0;
            for (int i = 0; i < encoded.length; i++) {
                sum += encoded[i];
            }
            
            //If there are even 1s, then we are good
            //If there are odd 1s, then there must be two errors which we cannot fix
            if (sum % 2 == 0) {
                
                //Now the sum is even, so we are goo.
                //Generate the data from the 8 digit encoded code
                //And return the data
                int[] decodedData = {encoded[2], encoded[4], encoded[5], encoded[6]};
                return decodedData;
            }
            else {
                
                //No we are not good.
                //Print the Error and return null.
                System.out.println("Error Occurs, May Contain 2 Errors.");
                return null;
            }
        }
        else {
            
            //Ok so there are no error in the first 7 bits.
            //And for now, it doesn't matter whether p8 is correct or not
            //So just return the data!
            return data;
        }
    }
    
    /**
     * Get the 3 parity bits of a length 4 data
     * @param data the original length 4 data
     * @return int[] the 3 parity bits
     */
    private static int[] getParity(int[] data) {
        int[] parity = {0, 0, 0};
        for (int i = 0; i < parity.length; i++) {
            int sum = 0;
            for (int j = 0; j < parityCoversDigit[i].length; j++) {
                sum += data[parityCoversDigit[i][j]];
            }
            parity[i] = sum % 2;
        }
        return parity;
    }
    
    /**
     * Change the int array to data string
     * @param data the data containing 0 and 1 ints.
     * @return string
     */
    private static String toString(int[] data) {
        String str = "";
        for (int i = 0; i < data.length; i++) {
            str += data[i];
        }
        return str;
    }
    
    /**
     * Change the string to int array
     * @param data to be translated
     */
    private static int[] toArray(String data) {
        int[] arr = new int[data.length()];
        for (int i = 0; i < data.length(); i++) {
            if (data.charAt(i) == '0') {
                arr[i] = 0;
            }
            else {
                arr[i] = 1;
            }
        }
        return arr;
    }
    
    /**
     * Print the data given the int array data
     * @param data the data to be printed
     */
    private static void printData(int[] data) {
        if (data != null) {
            System.out.println(toString(data));
        }
    }
    
    /**
     * Main method for testing Hamming
     * @param args
     */
    public static void main(String[] args) {
        System.out.println("\nTESTING 1011 --- Encoded: 01100110");
        printData(encode(toArray("1011")));
        printData(decode(toArray("01100110"))); //Correct One
        printData(decode(toArray("11100110"))); //Has One Error
        printData(decode(toArray("00100110"))); //Has One Error
        printData(decode(toArray("01000110"))); //Has One Error
        printData(decode(toArray("01110110"))); //Has One Error
        printData(decode(toArray("01101110"))); //Has One Error
        printData(decode(toArray("01100010"))); //Has One Error
        printData(decode(toArray("01100100"))); //Has One Error
        printData(decode(toArray("01100111"))); //Has One Error
        printData(decode(toArray("00000110"))); //Has Two Errors
        printData(decode(toArray("10100110"))); //Has Two Errors
        printData(decode(toArray("01100000"))); //Has Two Errors
        printData(decode(toArray("01100011"))); //Has Two Errors
        printData(decode(toArray("01000111"))); //Has Two Errors

        System.out.println("\nTESTING 0101 --- Encoded: 01001011");
        printData(encode(toArray("0101")));
        printData(decode(toArray("01001011")));
        printData(decode(toArray("11001011")));
        printData(decode(toArray("10001011")));
    }
}

File: hamming.js
var Hamming = {
    parityCoversDigit: [[0, 1, 3],  //p1覆盖d1, d2, d4
                        [0, 2, 3],  //p2覆盖d1, d3, d4
                        [1, 2, 3]], //p3覆盖d2, d3, d4
                        
    /**
     * Data should be a length 4 string: [d1, d2, d3, d4]
     * Return a length 8 encoded data
     */
    encode: function (data) {
        
        //计算每一个Parity Bit
        var parity = [0, 0, 0]; //[p1, p2, p3]先初始化为0
        for (var i = 0; i < parity.length; i++) {
            var sum = 0; //用于记录当前Parity Bit覆盖了多少个1
            for (var j = 0; j < this.parityCoversDigit[i].length; j++) {
                sum += parseInt(data[this.parityCoversDigit[i][j]]);
            }
            parity[i] = sum % 2; //如果覆盖了偶数个1，那么这个Parity Bit为0；如果奇数个，则为1
        }
        
        //将parity bits和data bits合起来，得到一个七位的新的data
        var ret = "" + parity[0] + parity[1] + data[0] + parity[2] + data[1] + data[2] + data[3];
        
        //开始给这个新的七位data附加一个最后的Parity Bit
        var sum = 0; //开始找这个七位data里面有多少个1
        for (var i = 0; i < ret.length; i++) {
            sum += parseInt(ret[i]);
        }
        ret += sum % 2; //如果共有偶数个，则这个最后附加的Parity Bit为0；如果奇数个，则为1
        
        //返回这个加密后的数据
        return ret;
    },
    
    /**
     * Encoded should be a length 8 encoded data.
     */
    decode: function (encoded) {
        
        //把必要的数据都好好地存下来
        var parity = [encoded[0], encoded[1], encoded[3]];
        var finalParity = encoded[7];
        var data = [encoded[2], encoded[4], encoded[5], encoded[6]];
        
        //设置一个正确的Parity
        var correctParity = [0, 0, 0]; //先初始化
        for (var i = 0; i < correctParity.length; i++) {
            var sum = 0;
            for (var j = 0; j < this.parityCoversDigit[i].length; j++) {
                sum += parseInt(data[this.parityCoversDigit[i][j]]);
            }
            correctParity[i] = sum % 2;
        }
        
        //比较正确的Parity和encoded中的parity是否相同
        var comparison = [false, false, false]; //初始化。相同即是true，不同即是false
        for (var i = 0; i < parity.length; i++) {
            comparison[i] = (parity[i] == correctParity[i]);
        }
        
        //开始根据比较的内容决定到底要修改data中的哪一位
        var incorrectDataBit = -1; //初始化为-1
        if (!comparison[0] && !comparison[1] && comparison[2]) { //p1和p2错误，p3正确
            incorrectDataBit = 1;
        }
        if (!comparison[0] && comparison[1] && !comparison[2]) { //p1和p3错误，p2正确
            incorrectDataBit = 0;
        }
        if (comparison[0] && !comparison[1] && !comparison[2]) { //p2和p3错误，p1正确
            incorrectDataBit = 2;
        }
        if (!comparison[0] && !comparison[1] && !comparison[2]) { //如果三个都错了
            incorrectDataBit = 3;
        }
        
        //如果有错，将那一位翻转，并检查此修复是否有错
        if (incorrectDataBit != -1) {
            data[incorrectDataBit] = (data[incorrectDataBit] == 0) ? 1 : 0;
            
            //统计新的数据里面共有多少个1
            var newEncoded = [parity[0], parity[1], data[0], parity[2], data[1], data[2], data[3], finalParity];
            var sum = 0;
            for (var i = 0; i < newEncoded.length; i++) {
                sum += parseInt(newEncoded[i]);
            }
            
            //如果是偶数个1，则正确，直接返回data
            //如果是奇数个1，则修改的不正确，弹出错误
            if (sum % 2 == 0) {
                return "" + data[0] + data[1] + data[2] + data[3];
            }
            else {
                console.log("Contains 2 Errors. Unable to recover");
                return "";
            }
        }
        else {
            
            //如果incorrectDataBit是-1，即表明没错，或，至少在data中没有错误。此时直接返回data就好
            return "" + data[0] + data[1] + data[2] + data[3];
        }
    }
}

console.log("\nTESTING 1011 --- Encoded: 01100110");
console.log(Hamming.encode("1011"));
console.log(Hamming.decode("01100110")); //Correct One
console.log(Hamming.decode("11100110")); //Has One Error
console.log(Hamming.decode("10100110")); //Has Two Errors

console.log("\nTESTING 0101 --- Encoded: 01001011");
console.log(Hamming.encode("0101"));
console.log(Hamming.decode("01001011"));
console.log(Hamming.decode("11001011"));
console.log(Hamming.decode("10001011"));
